ssm/variational.py

```diff
@@ -283,75 +334,138 @@ def sample_discrete_states(self):
                 for prms in self.params]
 
     def sample_continuous_states(self):
-        return [block_tridiagonal_sample(prms["J_diag"], prms["J_lower_diag"], prms["h"])
+        return [kalman_info_sample(prms["J_ini"], prms["h_ini"], 0,
+                                   prms["J_dyn_11"], prms["J_dyn_21"], prms["J_dyn_22"],
+                                   prms["h_dyn_1"], prms["h_dyn_2"], 0,
+                                   prms["J_obs"], prms["h_obs"], 0)
                 for prms in self.params]
 
+
+
     def sample(self):
         return list(zip(self.sample_discrete_states(), self.sample_continuous_states()))
 
     @property
     def mean_discrete_states(self):
         # Now compute the posterior expectations of z under q(z)
+        # NOTE: This returns the log normalizer, E[z_t], E[z_t, z_{t+1}]
+        # TODO: cache expectations and log normalizer here, so that they are
+        # available for computing entropies, but need to be sure that the parameters
+        # haven't changed.
         return [hmm_expected_states(prms["pi0"], prms["Ps"], prms["log_likes"])
                 for prms in self.params]
 
     @property
     def mean_continuous_states(self):
         # Now compute the posterior expectations of z under q(z)
-        return [block_tridiagonal_mean(prms["J_diag"], prms["J_lower_diag"], prms["h"], lower=True).reshape(np.shape(prms["h"]))
+        full_expectations = self.continuous_expectations
+        return [exp[1] for exp in full_expectations]
+
+    @property
+    def continuous_expectations(self):
+        return [kalman_info_smoother(prms["J_ini"], prms["h_ini"], 0,
```

Comment:
We can save some computation by caching the most recent results of `mean_continuous_states` and `mean_discrete_states` and the parameter values that were used to compute them.  Then we check if the parameters are the same and return the saved results if so.